### DISCUSSION TITLE
Case insensitive custom filter operation

### DISCUSSION THREAD

**Discussion by naufilshaikh09** [ChilliCream/hotchocolate#3915] [Show and tell]
Hey, if you want to add case insensitive custom filter operation in hotchocolate then you can find solution below.

For ex : 
```
query {
  users(where: { username: { contains: "admin" } }) {
    firstname
    lastname
  }
}
```

Solution :
```
protected override void Configure(IFilterConventionDescriptor descriptor)
{
    descriptor.AddDefaults();
    descriptor.Provider(
        new QueryableFilterProvider(
            x => x
                .AddDefaultFieldHandlers()
                .AddFieldHandler<QueryableStringInvariantContainsHandler>()));
}
```

```
public class QueryableStringInvariantContainsHandler : QueryableStringOperationHandler
{
    private static readonly MethodInfo _contains = typeof(string).GetMethod("IndexOf",
            new[] { typeof(string), typeof(StringComparison) });

    protected override int Operation => DefaultFilterOperations.Contains;
    public override Expression HandleOperation(QueryableFilterContext context, IFilterOperationField field, IValueNode value, object parsedValue)
    {
        Expression property = context.GetInstance();
        if (parsedValue is string str)
        {
            return Expression.NotEqual(
                    Expression.Call(
                        property,
                        _contains,
                        Expression.Constant(str, typeof(string)),
                        Expression.Constant(StringComparison.InvariantCultureIgnoreCase, typeof(StringComparison))
                    ),
                    Expression.Constant(-1, typeof(int))
                );
        }
        throw new InvalidOperationException();
    }
}
```

Add filter convention in Startup,cs
```
services.AddGraphQLServer()
                    .AddFiltering<CustomFilteringConvention>()
                    .AddConvention<IFilterConvention>(
                        new FilterConventionExtension(
                            x => x.AddProviderExtension(
                                new QueryableFilterProviderExtension(
                                    y => y.AddFieldHandler<QueryableStringInvariantContainsHandler>()))));
```

you can also refer this link too : [C# Dynamic lambda problem with implementing "IndexOf" ignore case](https://dynamic-linq.net/knowledge-base/57830194/csharp-dynamic-lambda-problem-with-implementing--indexof--ignore-case)

**Comment by lucas-garrido** [comment]
i got error "message: "'_s0.Name.IndexOf(\"test\", InvariantCultureIgnoreCase)' cannot be converted to SQL.""

**Reply by PascalSenn** [reply]
@lucas-garrido  What databse are. you using. It's likely that not every provider suppoorts invariantCultureIngoreCase

**Reply by lucas-garrido** [reply]
you are right i'm using linqToDb with hanadb provider... can i get same result with GetMethod("Contains") ?

**Reply by PascalSenn** [reply]
If you want case insensitivity you could also ToLower variable and column

**Reply by naufilshaikh09** [reply]
> i got error "message: "'_s0.Name.IndexOf("test", InvariantCultureIgnoreCase)' cannot be converted to SQL.""

Hey Lucas, I am using postgresql.

**Reply by lucas-garrido** [reply]
Do you have any example to provide ? that's what im trying to do but i dont know how i can to_lower on the column. 🙏
![image](https://user-images.githubusercontent.com/48165310/124760427-067a4500-df31-11eb-880e-e8162b0bb196.png)

**Reply by naufilshaikh09** [reply]
> you are right i'm using linqToDb with hanadb provider... can i get same result with GetMethod("Contains") ?

Please check this link. Hopefully it will resolve your isssue.
https://stackoverflow.com/questions/841226/case-insensitive-string-compare-in-linq-to-sql

**Reply by naufilshaikh09** [reply]
> Do you have any example to provide ? that's what im trying to do but i dont know how i can to_lower on the column. 🙏
> ![image](https://user-images.githubusercontent.com/48165310/124760427-067a4500-df31-11eb-880e-e8162b0bb196.png)

Look at that example. https://chillicream.com/docs/hotchocolate/api-reference/extending-filtering/#extending-iqueryable

**Reply by lucas-garrido** [reply]
Yes, that's what i was trying to do: make this example work with "Contains". But due to my poor skills i'm not succeeding😁i'm pretty new at csharp, any working example would be helpful

**Reply by PascalSenn** [reply]
@lucas-garrido  Expressions are a complex thing. You can think about an expression tree as you would write c# by code. 
When we compile a query, we compile expressions at runtime. The filter interceptor can intercept this compilation and add own parts to it. 

With this you get the reference to the Method IndexOf with the parameters string and StringComparison
```csharp
    private static readonly MethodInfo _indexOf = typeof(string).GetMethod("IndexOf",
            new[] { typeof(string), typeof(StringComparison) });
```
Lets say you now have something like `string.IndexOf`. what we need is `yourPorpertyName.IndexOf(str, StringComparison.InvariantCultureIgnoreCase)`

This is the bit 
```csharp
 Expression.Call(
    property,   // yourPropertyName.
    _indexOf, // IndexOf(
    Expression.Constant(str, typeof(string)), // str, 
    Expression.Constant(StringComparison.InvariantCultureIgnoreCase, typeof(StringComparison)) // StringComparison.InvariantCultureIgnoreCase)
                    ),
```

**Reply by lucas-garrido** [reply]
@PascalSenn @naufil09 Thank you guys, it works ! 🥳

**Reply by kbiefang** [reply]
Hey, I am having a similar problem. Im getting the following error: 

> "Additional information: Translation of method 'string.IndexOf' failed.

I am using Postgres 11.12 with EFCore and our resolvers return an IQueryable. As far as I understand LINQ to Entities is responsible for storing our statement in the DB to allow for deferred execution. As I read [here](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/ef/language-reference/clr-method-to-canonical-function-mapping?redirectedfrom=MSDN#systemstring-method-instance-mapping) the used overload IndexOf(string, StringComparison) is not canonical. Why is this working for you, @naufil09 but not for me? 

In another [Stackoverflow post](https://stackoverflow.com/questions/23211725/linq-to-entities-does-not-recognize-the-method-int32-indexofsystem-string-sys) I have read something about changing to collation of the database. Might this help?

**Reply by naufilshaikh09** [reply]
Hey @kbiefang, sorry for late reply.

Your problem is resolved or facing same problem ?

I don't know why it isn't working for you, can you please explain more ?
 You can follow this link https://chillicream.com/docs/hotchocolate/api-reference/extending-filtering/#extending-iqueryable or ask question in slack channel too https://hotchocolategraphql.slack.com/. 

I am having still little problem in method. With data it's working properly but if there is no data then it is not working properly.

Example :
![image](https://user-images.githubusercontent.com/62197059/129486150-d6a0f0bd-1461-41a5-ac58-b9afdda4f71c.png)

**Reply by kbiefang** [reply]
Hey,

as I said I think LINQ to Entities is unable to store the IndexOf(string, StringComparison) operation, because there is no equivalent in the database implementation. 

I worked around this by rewriting the expression to use ToLower() and IndexOf(string), which might not be the best solution, but for now it is working.

**Comment by arjundnair-fingent** [comment]
@naufilshaikh09 This works for me, but not for the inner collections. 

For example, my class is 
```
public class User
{
    public string[] Roles { get; set; }

    public IEnumerable<Address> Addresses { get; set; }
}
```

```
public class Address
{
    public string Street { get; set; }

    public bool IsPrimary { get; set; }
}
```

And I need to apply case-insensitive filter on the `Addresses` object. In that case this solution does not work. Can you please help me by finding a solution for that? Thanks in advance.

Note: My HotChocolate version is V11.

Regards,
Arjun

**Reply by europanda** [reply]
Work for me with PostgreSQL. HotChocolate v12.4.1


```
    public class QueryableStringInvariantContainsHandler : QueryableStringOperationHandler
    {
        public QueryableStringInvariantContainsHandler(InputParser inputParser) : base(inputParser) { }

        protected override int Operation => DefaultFilterOperations.Contains;
        public override Expression HandleOperation(QueryableFilterContext context, IFilterOperationField field, IValueNode value, object parsedValue)
        {
            Expression property = context.GetInstance();
            if (parsedValue is string str)
            {
                var toLower = Expression.Call(property, typeof(string).GetMethod("ToLower", Type.EmptyTypes));
                return Expression.Call(toLower, typeof(string).GetMethod(nameof(string.Contains), new Type[] { typeof(string) }), Expression.Constant(str.ToLower()));
            }
            throw new InvalidOperationException();
        }
```

in Startup.cs

```
                .AddConvention<IFilterConvention>(new 
                FilterConventionExtension(x => x.AddProviderExtension(
                    new QueryableFilterProviderExtension(y => y.AddFieldHandler<QueryableStringInvariantContainsHandler>()))));
```

**Reply by arjundnair-fingent** [reply]
Thanks @europanda. I'll try this out tomorrow and let you know the result.

**Reply by arjundnair-fingent** [reply]
@europanda I'll explain the scenario. Please help me if you get some solution. Please note that I'm using SQL Server 2019 v15 with HotChocolate v11.3.5

I have my class like this: 

```
public class User
{
    public string Name { get; set; }

    [UseFiltering]
    public IEnumerable<Address> Addresses { get; set; }
}

public class Address
{
    public string Street { get; set; }

    public bool IsPrimary { get; set; }
}
```


Also I have the resolver for `User` class like this 
```
[UseOffsetPaging]
[UseProjection]
[UseFiltering]
[UseSorting]
public IQueryable<User> GetUsers([Service] IUserService userService)
{
    return userService.GetUsers();
}

```

I tried the query: 
```
query{
  users(skip: 0, take: 4){
    items{
      name,
      addresses(where: { street: {contains: "baker"} })
      {
        street,
        isPrimary
      }
    }
  }
}

```

Each user may have more than one addresses. My requirement is that when I'm searching for **baker**, it should only show the address which contains the string **baker** in it, eg: "221 Baker Street, London". If there is no address with the string **baker** in it, then the address collection must be empty. It should also be case-insensitive. I tried the code you've mentioned above, but it doesn't work. I hope my requirement is clear for you.

**Reply by europanda** [reply]
Try this query:
```
query
{
users(where: {addresses: {some: {street: {contains: "baker"}}}})
 {
  items
  {
    name
    addresses
    {
      street
      isPrimary
    }
  }
 }
}
``
![Query](https://user-images.githubusercontent.com/25351137/153588179-f481abf5-b556-4832-96d6-6f2b06360bb9.PNG)
`

**Reply by europanda** [reply]
A fully working and tested project  [https://github.com/europanda/GraphQL-Example](https://github.com/europanda/GraphQL-Example)
Сonfigure the connection string in appsettings.json and run.

**Reply by lev-smplt** [reply]
@europanda Thank you for your solution. But I noticed that this code
```
public class QueryableStringInvariantContainsHandler : QueryableStringOperationHandler
    {
        public QueryableStringInvariantContainsHandler(InputParser inputParser) : base(inputParser) { }

        protected override int Operation => DefaultFilterOperations.Contains;
        public override Expression HandleOperation(QueryableFilterContext context, IFilterOperationField field, IValueNode value, object parsedValue)
        {
            Expression property = context.GetInstance();
            if (parsedValue is string str)
            {
                var toLower = Expression.Call(property, typeof(string).GetMethod("ToLower", Type.EmptyTypes));
                return Expression.Call(toLower, typeof(string).GetMethod(nameof(string.Contains), new Type[] { typeof(string) }), Expression.Constant(str.ToLower()));
            }
            throw new InvalidOperationException();
        }
  }
```
will generate sql (for postgresql) like this:
`... strpos(lower(a.Name), 'requested_string')>0`
As I know `strpos` lacks the ability to leverage indexes on `a.Name` and requires scanning the entire table and performing the substring search for each row. It would be better to generate something like this: `a.Name ilike '%requested_string%'`.
I found problem but not solution, have not got yet how to force generate sql with `ilike`.

**Reply by ciprian-aw** [reply]
@lev-smplt this is what worked for me:

```
public class QueryableStringInvariantContainsHandler : QueryableStringOperationHandler
{
    private static readonly Regex sWhitespace = new Regex(@"\s+");
    private static readonly MethodInfo _ilike = typeof(NpgsqlDbFunctionsExtensions).GetMethod("ILike",
            new[] { typeof(DbFunctions), typeof(string), typeof(string) });

    public QueryableStringInvariantContainsHandler(InputParser inputParser)
        : base(inputParser)
    {
    }

    protected override int Operation => DefaultFilterOperations.Contains;

    public override Expression HandleOperation(
        QueryableFilterContext context,
        IFilterOperationField field,
        IValueNode value,
        object parsedValue)
    {
        Expression property = context.GetInstance();
        var source = Expression.Constant(EF.Functions);
        if (parsedValue is string str)
        {
            var pattern = sWhitespace.Replace(parsedValue as string, "%");
            pattern = $"%{pattern}%";

            return Expression.AndAlso(
                Expression.NotEqual(property, Expression.Constant(null, typeof(object))),
                Expression.Call(
                    null,
                    _ilike,
                    source,
                    property,
                    // Expression.Constant(pattern)
                    CreateParameter(pattern, property.Type)
                )
            );
        }
        throw new InvalidOperationException();
    }

    private static Expression CreateAndConvertParameter<T>(object value)
    {
        Expression<Func<T>> lambda = () => (T)value;
        return lambda.Body;
    }

    private static Expression CreateParameter(object? value, Type type)
        => (Expression)_createAndConvert
            .MakeGenericMethod(type)
            .Invoke(null,
                new[] { value })!;

    private static readonly MethodInfo _createAndConvert =
        typeof(FilterExpressionBuilder)
            .GetMethod(nameof(CreateAndConvertParameter), NonPublic | Static)!;
}
```

**Reply by lev-smplt** [reply]
@ciprian-aw, thanks for reply and for solution :) I'll try it.

**Comment by arjundnair-fingent** [comment]
@europanda This is not what I seek. This is the query to filter the parent `Users` with the `Addresses` child. I just want to filter the `Addresses` child without affecting the `Users` parent.

**Comment by arjundnair-fingent** [comment]
@europanda My friend has found out another solution. Thanks for sparing some of your time for me.

**Reply by jackdh** [reply]
Can you share that solution @arjundnair-fingent ?

**Comment by VILLAN3LL3** [comment]
> Hey, if you want to add case insensitive custom filter operation in hotchocolate then you can find solution below.
> 
> For ex :
> 
> ```
> query {
>   users(where: { username: { contains: "admin" } }) {
>     firstname
>     lastname
>   }
> }
> ```
> 
> Solution :
> 
> ```
> protected override void Configure(IFilterConventionDescriptor descriptor)
> {
>     descriptor.AddDefaults();
>     descriptor.Provider(
>         new QueryableFilterProvider(
>             x => x
>                 .AddDefaultFieldHandlers()
>                 .AddFieldHandler<QueryableStringInvariantContainsHandler>()));
> }
> ```
> 
> ```
> public class QueryableStringInvariantContainsHandler : QueryableStringOperationHandler
> {
>     private static readonly MethodInfo _contains = typeof(string).GetMethod("IndexOf",
>             new[] { typeof(string), typeof(StringComparison) });
> 
>     protected override int Operation => DefaultFilterOperations.Contains;
>     public override Expression HandleOperation(QueryableFilterContext context, IFilterOperationField field, IValueNode value, object parsedValue)
>     {
>         Expression property = context.GetInstance();
>         if (parsedValue is string str)
>         {
>             return Expression.NotEqual(
>                     Expression.Call(
>                         property,
>                         _contains,
>                         Expression.Constant(str, typeof(string)),
>                         Expression.Constant(StringComparison.InvariantCultureIgnoreCase, typeof(StringComparison))
>                     ),
>                     Expression.Constant(-1, typeof(int))
>                 );
>         }
>         throw new InvalidOperationException();
>     }
> }
> ```
> 
> Add filter convention in Startup,cs
> 
> ```
> services.AddGraphQLServer()
>                     .AddFiltering<CustomFilteringConvention>()
>                     .AddConvention<IFilterConvention>(
>                         new FilterConventionExtension(
>                             x => x.AddProviderExtension(
>                                 new QueryableFilterProviderExtension(
>                                     y => y.AddFieldHandler<QueryableStringInvariantContainsHandler>()))));
> ```
> 
> you can also refer this link too : [C# Dynamic lambda problem with implementing "IndexOf" ignore case](https://dynamic-linq.net/knowledge-base/57830194/csharp-dynamic-lambda-problem-with-implementing--indexof--ignore-case)

This doesn't seem to work if there are null values in the database. Any suggestions?

**Reply by VILLAN3LL3** [reply]
Change the `HandleOperation` expression as follows to make it work with null values in the database:
```
public override Expression HandleOperation(
            QueryableFilterContext context,
            IFilterOperationField field,
            IValueNode value,
            object parsedValue)
        {
            Expression property = context.GetInstance();
            if (parsedValue is string str)
            {
                return Expression.AndAlso(
                    Expression.NotEqual(property, Expression.Constant(null, typeof(object))),
                    Expression.NotEqual(
                        Expression.Call(
                            property,
                            typeof(string).GetMethod("IndexOf", new[] { typeof(string), typeof(StringComparison) }),
                            Expression.Constant(str),
                            Expression.Constant(StringComparison.OrdinalIgnoreCase)
                        ),
                        Expression.Constant(-1)
                    )
                );
            }

            throw new InvalidOperationException();
        }
```

**Reply by VILLAN3LL3** [reply]
If you want to also ignore accents, you can use the following code (caballé == caballe, Fuß == fuss):
```
public override Expression HandleOperation(
            QueryableFilterContext context,
            IFilterOperationField field,
            IValueNode value,
            object parsedValue)
        {
            Expression property = context.GetInstance();
            if (parsedValue is string str)
            {
                // Bad way of doing it but Using ILIKE on DB level is somehow not working.
                // https://github.com/npgsql/efcore.pg/issues/618
                return Expression.AndAlso(
                    Expression.NotEqual(property, Expression.Constant(null, typeof(object))),
                    Expression.NotEqual(
                        Expression.Call(
                            Expression.Constant(CultureInfo.InvariantCulture.CompareInfo),
                            typeof(CompareInfo).GetMethod(nameof(CompareInfo.IndexOf), [typeof(string), typeof(string), typeof(CompareOptions)]),
                            property,
                            Expression.Constant(str),
                            Expression.Constant(CompareOptions.IgnoreNonSpace | CompareOptions.IgnoreCase)
                        ),
                        Expression.Constant(-1)
                    )
                );
            }

            throw new InvalidOperationException();
        }
```

**Comment by martinRocks** [comment]
I love this project, however, using HotChocolate.AspNetCore version 16.0.2, the documentation located here (https://chillicream.com/docs/hotchocolate/v16/fetching-data/integrations/extending-filtering/) does not compile.  I have changed the code to be like below, but it just doesn't work.  Help.

```public sealed class CustomFilteringConvention : FilterConvention
{
    protected override void Configure(IFilterConventionDescriptor descriptor)
    {
        descriptor.AddDefaults();
        descriptor.Provider(
            new QueryableFilterProvider(
                x => x
                    .AddDefaultFieldHandlers()
                    .AddFieldHandler(y => new QueryableStringInvariantEqualsHandler(y.InputParser))));
    }
}

**Reply by tobias-tengler** [reply]
Could you please open an issue for this? That way we can track it better. I'll have a look already in the meantime.

**Reply by martinRocks** [reply]
I will do it right now.  Thank you
